provider/aws/aws.go

```diff
@@ -494,9 +497,30 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
 	for z, cs := range changesByZone {
 		var failedUpdate bool
 
-		batchCs := batchChangeSet(cs, p.batchChangeSize)
+		// group changes into new changes and into changes that failed in a previous iteration and are retried
+		var newChanges, retriedChanges []*route53.Change
+		for _, c := range cs {
+			found := false
+			if p.failedChangesQueue[z] != nil {
+				for _, failedChange := range p.failedChangesQueue[z] {
+					if c == failedChange {
+						retriedChanges = append(retriedChanges, c)
+						found = true
+					}
+				}
+				p.failedChangesQueue[z] = nil // clear the queue
```

Comment:
Why are we doing this loop if we are going to append them anyway in line 517?

Reply:
The purpose of the loop is to sort them changes into new changes / changes that have failed last time. We clear the list here because the change might succeed this time (if it doesn't, it gets added back to the list below in the error handling). On line 517 the actual "sorting" is happening in a sense that new changes are tried first (because they have a higher likelihood of succeeding), and retried changes are coming after the new changes.

Reply:
Do I understand it right that this has no impact on correctness? Does this likelihood have any sort of impact on actual usage? If not, can we remove it and simplify the code?

Reply:
As far as I can tell it has no influence on the correctness, however it reduces the overall amount of API calls (which can help prevent API rate limiting) because new changes have a higher likelihood of succeeding - they are tried first in a separate transaction, which is unlikely to fail. If they would be randomly mixed with previously failing changes, on *every* new change, you would have additional API calls (complete batch with new+old changes fails, then all changes are tried one by one). So if you have e.g. 4 failing changes and 4 new ones, you would do 9 calls (on batch with all 8 changes that fails + one batch with each change for retry) instead of only 5 (one batch with the 4 new changes + one batch with each failing change for retry).

Reply:
I like the idea of trying to handle the new changes in their own transaction first (pretty much mimicking the "happy path" as if we didn't have any failed changes in the queue) because there are likely to succeed. The users of these records wouldn't be impacted by any delay due to the handling of failed batches (due to splitting and more api calls).

Reply:
@alfredkrohmer How much more complicated is the code because of that in your opinion?

Reply:
Oh I see, it's basically just this part here where we separate the `failedChanges` from the `desiredChanges` in order to move the remaining (`newChanges`) to the front of the queue.

If that's true we could make this more obvious for the reader with comments for some meaningful function names.